jni/src/commons.cpp

```diff
@@ -51,6 +51,26 @@ jlong knn_jni::commons::storeByteVectorData(knn_jni::JNIUtilInterface *jniUtil,
         vect->clear();
     }
 
+    int dim = jniUtil->GetInnerDimensionOf2dJavaByteArray(env, dataJ);
+    jniUtil->Convert2dJavaObjectArrayAndStoreToBinaryVector(env, dataJ, dim, vect);
+
+    return (jlong) vect;
+}
+
+jlong knn_jni::commons::storeByteVectorData(knn_jni::JNIUtilInterface *jniUtil, JNIEnv *env, jlong memoryAddressJ,
+                                        jobjectArray dataJ, jlong initialCapacityJ, jboolean appendJ) {
+    std::vector<int8_t> *vect;
+    if (memoryAddressJ == 0) {
+        vect = new std::vector<int8_t>();
+        vect->reserve(static_cast<long>(initialCapacityJ));
+    } else {
+        vect = reinterpret_cast<std::vector<int8_t>*>(memoryAddressJ);
+    }
+
+    if (appendJ == JNI_FALSE) {
+            vect->clear();
```

Comment:
The clear operation doesn't come for free and it runs in linear complexity.

Ref: https://cplusplus.com/reference/vector/vector/clear/ I think this append feature will add latency in the whole streaming operation.

cc: @shatejas

Reply:
We are trading a memory optimization for latencies here. If the user doesn't want to rely on underlying implementation optimization and be absolutely sure that this doesn't incur latencies, the can always use append == true which is still the default in [JNICommons](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/jni/JNICommons.java#L48).

This was done to have no fragmentation for iterative builds. Alternative is to allow memory fragmentation, considering indexing is not latency sensitive and is memory sensitive, it makes sense to keep it. Thoughts?

I also dug into 
> This may be optimized to constant complexity for [trivially-destructible types](https://cplusplus.com/is_trivially_destructible) (such as scalar or PODs), where elements need not be destroyed.

Here is a [good explanation](https://stackoverflow.com/questions/11235975/what-is-the-complexity-of-stdvectortclear-when-t-is-a-primitive-type#:~:text=Therefore%2C%20setting%20vector%20's%20size,by%20employing%20partial%20template%20specialization) of why it might possible be O(1) for primitive types considering they are trivially destructible. So we might possibly get away with added latencies but cannot say for sure

Reply:
I think this operation is fine to have - int8_t should be trivially destructable. In future, we should move away from using vector so we can more explicitly control the allocation for the buffer.

Reply:
If we are confident on the destruction that I agree it should not be a big performance concern. Resolving the conversation.